conformance/tests/httproute-weight.go

```diff
@@ -0,0 +1,157 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tests
+
+import (
+	"cmp"
+	"errors"
+	"fmt"
+	"math"
+	"slices"
+	"strings"
+	"sync"
+	"testing"
+
+	"golang.org/x/sync/errgroup"
+	"k8s.io/apimachinery/pkg/types"
+
+	"sigs.k8s.io/gateway-api/conformance/utils/http"
+	"sigs.k8s.io/gateway-api/conformance/utils/kubernetes"
+	"sigs.k8s.io/gateway-api/conformance/utils/suite"
+)
+
+func init() {
+	ConformanceTests = append(ConformanceTests, HTTPRouteWeight)
+}
+
+var HTTPRouteWeight = suite.ConformanceTest{
+	ShortName:   "HTTPRouteWeight",
+	Description: "An HTTPRoute with weighted backends",
+	Manifests:   []string{"tests/httproute-weight.yaml"},
+	Features: []suite.SupportedFeature{
+		suite.SupportGateway,
+		suite.SupportHTTPRoute,
+	},
+	Test: func(t *testing.T, suite *suite.ConformanceTestSuite) {
+		var (
+			ns      = "gateway-conformance-infra"
+			routeNN = types.NamespacedName{Name: "weighted-backends", Namespace: ns}
+			gwNN    = types.NamespacedName{Name: "same-namespace", Namespace: ns}
+			gwAddr  = kubernetes.GatewayAndHTTPRoutesMustBeAccepted(t, suite.Client, suite.TimeoutConfig, suite.ControllerName, kubernetes.NewGatewayRef(gwNN), routeNN)
+		)
+
+		kubernetes.HTTPRouteMustHaveResolvedRefsConditionsTrue(t, suite.Client, suite.TimeoutConfig, routeNN, gwNN)
+
+		t.Run("Requests should have a distribution that matches the weight", func(t *testing.T) {
+			expected := http.ExpectedResponse{
+				Request:   http.Request{Path: "/"},
+				Response:  http.Response{StatusCode: 200},
+				Namespace: "gateway-conformance-infra",
+			}
+
+			// Assert request succeeds before doing our distribution check
+			http.MakeRequestAndExpectEventuallyConsistentResponse(t, suite.RoundTripper, suite.TimeoutConfig, gwAddr, expected)
+
+			for i := 0; i < 10; i++ {
+				if err := testDistribution(t, suite, gwAddr, expected); err != nil {
+					t.Logf("Traffic distribution test failed (%d/10): %s", i+1, err)
+				} else {
+					return
+				}
+			}
```

Comment:
Nit: I think it would be better if this was a cumulative total. For example, if we were targeting an equal distribution between backends, and the first iteration send 40% to A and 60% to B while the second flipped those (60% to A and 40% to A), I think we'd want that test to pass.

Reply:
Interesting. Why? Do implementations commonly implement weights in a cumulative way? 

Like I think if one iteration got 10% and 90% and the other one got 90% and 10% not sure we are fine with that.